polkadot/node/network/approval-distribution/src/lib.rs

```diff
@@ -578,32 +772,100 @@ impl State {
 
 		adjust_required_routing_and_propagate(
 			ctx,
-			&self.peer_data,
 			&mut self.blocks,
 			&self.topologies,
 			|block_entry| block_entry.session == session,
 			|required_routing, local, validator_index| {
-				if *required_routing == RequiredRouting::PendingTopology {
-					*required_routing = topology
+				if required_routing == &RequiredRouting::PendingTopology {
+					topology
 						.local_grid_neighbors()
-						.required_routing_by_index(*validator_index, local);
+						.required_routing_by_index(*validator_index, local)
+				} else {
+					*required_routing
 				}
 			},
+			&self.peer_views,
 		)
 		.await;
 	}
 
+	async fn process_incoming_assignments<Context, R>(
+		&mut self,
+		ctx: &mut Context,
+		metrics: &Metrics,
+		peer_id: PeerId,
+		assignments: Vec<(IndirectAssignmentCertV2, CandidateBitfield)>,
+		rng: &mut R,
+	) where
+		R: CryptoRng + Rng,
+	{
+		for (assignment, claimed_indices) in assignments {
+			if let Some(pending) = self.pending_known.get_mut(&assignment.block_hash) {
+				let block_hash = &assignment.block_hash;
+				let validator_index = assignment.validator;
+
+				gum::trace!(
+					target: LOG_TARGET,
+					%peer_id,
+					?block_hash,
+					?claimed_indices,
+					?validator_index,
+					"Pending assignment",
+				);
+
+				pending.push((peer_id, PendingMessage::Assignment(assignment, claimed_indices)));
+
+				continue
+			}
+
+			self.import_and_circulate_assignment(
+				ctx,
+				metrics,
+				MessageSource::Peer(peer_id),
+				assignment,
+				claimed_indices,
+				rng,
+			)
+			.await;
+		}
+	}
+
 	async fn process_incoming_peer_message<Context, R>(
 		&mut self,
 		ctx: &mut Context,
 		metrics: &Metrics,
 		peer_id: PeerId,
-		msg: net_protocol::ApprovalDistributionMessage,
+		msg: Versioned<
+			protocol_v1::ApprovalDistributionMessage,
+			protocol_v2::ApprovalDistributionMessage,
+			protocol_vstaging::ApprovalDistributionMessage,
```

Comment:
So:
- v1 -  is the initial validation protocol.
- v2 - is the protocol when we added async backing, which we already released, so there is not way we can reuse the same version for this changes, because to offer backward compatibility we need to be aware of the situation where we have peers, that do not know the new assignment format, so we go and translate it for them.

- vstaging - is the protocol where the new Assignment format is introduced and it is meant to become V3, I didn't not make it V3 at this point because it is not meant to be activated on nodes until we get the Approval changes as well.

So, my idea was that we keep vstaging around guarded under https://github.com/paritytech/polkadot-sdk/pull/1178/files#diff-ccf6dac7c0f390c65d31b155e518051af3dea517b1722905ded32c5d02b88849R126, until https://github.com/paritytech/polkadot-sdk/pull/1191 lands as well, then we activate vstaging and make it V3. 

Otherwise, we are going to need a V4 for https://github.com/paritytech/polkadot-sdk/pull/1191 as well, which in the end is probably what will have to do if that takes long enough to get merge(but I hope it won't).

@sandreim, @eskimor let me know if you agree with the approach.

Reply:
Decided offline that this is going be the approach.